Add indexes for CASCADE deletes for task_instance

dstandish opened the pull request:
When we add foreign keys with ON DELETE CASCADE, and we delete rows in the foreign table, the database needs to join back to the referencing table.  If there's no suitable index, then it can be slow to perform the deletes.

cc @wolfier 

closes #24484

jedcunningham commented on airflow/migrations/versions/0111_2_4_0_add_indexes_for_cascade_deletes.py:
```diff
@@ -0,0 +1,68 @@
+down_revision = '3c94c427fdf6'
+branch_labels = None
+depends_on = None
+airflow_version = '2.4.0'
```
```suggestion
airflow_version = '2.3.3'
```

We should pull this into the next release.

jedcunningham commented on airflow/migrations/versions/0111_2_4_0_add_indexes_for_cascade_deletes.py:
```diff
@@ -0,0 +1,68 @@
+
+def upgrade():
+    """Apply Add indexes for CASCADE deletes"""
+    # ### commands auto generated by Alembic - please adjust! ###
```
```suggestion
```

nit

jedcunningham commented on airflow/migrations/versions/0111_2_4_0_add_indexes_for_cascade_deletes.py:
```diff
@@ -0,0 +1,68 @@
+            'idx_xcom_task_instance', ['dag_id', 'task_id', 'run_id', 'map_index'], unique=False
+        )
+
+    # ### end Alembic commands ###
```
```suggestion
```

jedcunningham commented on airflow/migrations/versions/0111_2_4_0_add_indexes_for_cascade_deletes.py:
```diff
@@ -0,0 +1,68 @@
+
+def downgrade():
+    """Unapply Add indexes for CASCADE deletes"""
+    # ### commands auto generated by Alembic - please adjust! ###
```
```suggestion
```

jedcunningham commented on airflow/migrations/versions/0111_2_4_0_add_indexes_for_cascade_deletes.py:
```diff
@@ -0,0 +1,68 @@
+    with op.batch_alter_table('task_fail', schema=None) as batch_op:
+        batch_op.drop_index('idx_task_fail_task_instance')
+
+    # ### end Alembic commands ###
```
```suggestion
```

jedcunningham commented on airflow/models/taskfail.py:
```diff
@@ -39,6 +39,7 @@ class TaskFail(Base):
     duration = Column(Integer)
 
     __table_args__ = (
+        Index("idx_task_fail_task_instance", dag_id, task_id, run_id, map_index, unique=False),
```
```suggestion
        Index("idx_task_fail_task_instance", dag_id, task_id, run_id, map_index),
```

That's the default, no?

jedcunningham commented on airflow/models/taskreschedule.py:
```diff
@@ -61,6 +61,7 @@ class TaskReschedule(Base):
             name="task_reschedule_ti_fkey",
             ondelete="CASCADE",
         ),
+        Index('idx_task_reschedule_dag_run', dag_id, run_id, unique=False),
```
```suggestion
        Index('idx_task_reschedule_dag_run', dag_id, run_id),
```

jedcunningham commented on airflow/migrations/versions/0111_2_4_0_add_indexes_for_cascade_deletes.py:
```diff
@@ -0,0 +1,68 @@
+    # ### commands auto generated by Alembic - please adjust! ###
+    with op.batch_alter_table('task_fail', schema=None) as batch_op:
+        batch_op.create_index(
+            'idx_task_fail_task_instance', ['dag_id', 'task_id', 'run_id', 'map_index'], unique=False
```
```suggestion
            'idx_task_fail_task_instance', ['dag_id', 'task_id', 'run_id', 'map_index']
```

Same here.

jedcunningham commented on airflow/migrations/versions/0111_2_4_0_add_indexes_for_cascade_deletes.py:
```diff
@@ -0,0 +1,68 @@
+        )
+
+    with op.batch_alter_table('task_reschedule', schema=None) as batch_op:
+        batch_op.create_index('idx_task_reschedule_dag_run', ['dag_id', 'run_id'], unique=False)
```
```suggestion
        batch_op.create_index('idx_task_reschedule_dag_run', ['dag_id', 'run_id'])
```

jedcunningham commented on airflow/migrations/versions/0111_2_4_0_add_indexes_for_cascade_deletes.py:
```diff
@@ -0,0 +1,68 @@
+
+    with op.batch_alter_table('xcom', schema=None) as batch_op:
+        batch_op.create_index(
+            'idx_xcom_task_instance', ['dag_id', 'task_id', 'run_id', 'map_index'], unique=False
```
```suggestion
            'idx_xcom_task_instance', ['dag_id', 'task_id', 'run_id', 'map_index']
```

dstandish committed:
Add indexes for CASCADE deletes

When we add foreign keys with ON DELETE CASCADE, and we delete rows in the foreign table, the database needs to join back to the referencing table.  If there's no suitable index, then it can be slow to perform the deletes.

dstandish committed:
skip index for mysql

dstandish committed:
handle mysql

dstandish committed:
fix offline

dstandish committed:
better descriptino
